### PR TITLE
Add target_encoder.R code part and basic tests

### LINKED ISSUE #26: Code target_encoder
This is just an issue to track the progress of the entire function (since the milestone instructions state to create an issue for each function, and we only created issues for each docstring in particular).

- [x] Docstring
- [x] Code
- [x] Tests

Will modify as we understand what we need to do in the coming weeks.

### LINKED ISSUE #32: Code test_target_encoder
This is just an issue to track the progress of the entire unit test function for #26

- [x]  Comments explaining each test.

- [x] 2 to 3 edge case tests

### PR DESCRIPTION
**Pull request by fsywang**
This pull request addresses #26 #32 #42 . The docstring has been modified.

### CODE REVIEW

File: R/target_encoder.R [hunk 1]
```diff
@@ -26,9 +30,111 @@
...
+  if (!objective %in% c("regression", "binary")) {
+    stop("Objective must be regression or binary.")
+  }
+  # check type of input of prior
+  if (!is.numeric(prior)) {
+    stop("The prior input should be numeric.")
+  }
+  # check type of input X_train
+  if (!is.data.frame(X_train)) {
+    stop("Type of X_train must be pd.Dataframe.")
```
Comment (braydentang1) [inline R/target_encoder.R]: There's a reference to a Pandas dataframe here.

File: R/target_encoder.R [hunk 2]
```diff
@@ -26,9 +30,111 @@
...
+      search_table['encodings'] <- (search_table['the_sum'] + prior * global_mean) / (search_table['the_count'] + prior)
+      search_table = search_table %>% dplyr::select( -c(the_sum,the_count))
+      # # encode categorical columns for training dataset
+      train_processed <- left_join(train_processed, search_table, by = column)
+      train_processed[column] <- train_processed['encodings']
+      train_processed <- train_processed %>% dplyr::select(-encodings)
+      }
+    out <- train_processed
+
+    } else{
```
Comment (braydentang1) [inline R/target_encoder.R]:
```suggestion
    } else {
```

just for tidyverse standards, that's all.

File: R/target_encoder.R [hunk 3]
```diff
@@ -26,9 +30,111 @@
...
+    }
+
+  }
+
+  train_processed <- X_train
+
+  global_mean <- mean(y_new, na.rm = TRUE)
+
+  if (is.null(X_test)){
+    for(i in seq_along(cat_columns)){
```
Comment (braydentang1) [inline R/target_encoder.R]:
```suggestion
  if (is.null(X_test)) {
    for (i in seq_along(cat_columns)) {
```

File: R/target_encoder.R [hunk 4]
```diff
@@ -26,9 +30,111 @@
...
+
+   y_new <- y
+
+  # binary objective
+  if (objective == 'binary') {
+    if (y %>% unique() %>% length() != 2) {
+      stop("The target variable must be binary.")
+    }
+    # encode target variable to 1 and 0
+    if (!is_double(y)){
```
Comment (braydentang1) [inline R/target_encoder.R]:
```suggestion
    if (!is_double(y)) {
```

File: R/target_encoder.R [hunk 5]
```diff
@@ -26,9 +30,111 @@
...
+      }
+
+      # check if X_test contains cat_columns
+      if (any(cat_columns %in% colnames(X_test)) == FALSE) {
+        stop("Column does not exist in the training set.")
+      }
+
+      test_processed <- X_test
+
+      for(i in seq_along(cat_columns)){
```
Comment (braydentang1) [inline R/target_encoder.R]:
```suggestion
      for (i in seq_along(cat_columns)) {
```

**Review by braydentang1 — CHANGES_REQUESTED** [review]

**Review by braydentang1 — APPROVED** [review]
